File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void solve(int n, int open, int close, string res, vector<string> &ans){

        if(open == close && open == n){
            ans.push_back(res);
            return;
        }

        // when no. of open bracket is less than n, then we can push open bracket in res
        if(open < n)
            solve(n, open + 1, close, res + '(', ans);
        
        // when no. of close bracket is less than open,then we can push open bracket in res
        if(close < open)
            solve(n, open, close + 1, res + ')', ans);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        // n, openBracket count, closeBracket count, tempString, ans
        solve(n, 0, 0,"", ans);
        return ans;
        
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        solve(n, 0, 0, new StringBuilder(), ans);
        return ans;
    }

    public void solve(int n, int open, int close, StringBuilder sb, List<String> ans){

        if(sb.length() == 2*n){
            ans.add(sb.toString());
            return;
        }

        if(open < n){
            sb.append('(');
            solve(n, open + 1, close, sb, ans);
            sb.deleteCharAt(sb.length() - 1);
        }

        if(close < open){
            sb.append(')');
            solve(n, open, close + 1, sb, ans);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}